### ISSUE TITLE
numpy.ndarray.__repr__ shorts incorrect/inconsistent dtype for unions

### ISSUE THREAD

**Issue by gerritholl** [numpy/numpy#6802]
For a structured `ndarray` with a union-style dtype (such as constructed with `(base_dtype, new_dtype)`), the `dtype` is incorrectly shown by `ndarray.__repr__`, as illustrated below:

```
In [948]: M = zeros(4, [("A", ">u4", [("comps", '2>u2')])])

In [949]: M
Out[949]:
array([(0,), (0,), (0,), (0,)],
      dtype=[('A', [('comps', '>u2', (2,))])])

In [950]: M["A"]
Out[950]: array([0, 0, 0, 0], dtype=uint32)

In [951]: M["A"].dtype
Out[951]: dtype([('comps', '>u2', (2,))])
```

Somehow both are correct as it has multiple dtypes if I understand things correctly (I'm new to unions), but it looks wrong to me to suggest that the dtype of `M["A"]` is just `uint32`.

**Comment by ahaldane** [comment]
Yes, the repr of such dtypes could use some fixing up. The `base_dtype` is often hidden. I already tried to fix this a little earlier this year in #5483, but only took case of void `base_dtype`. I don't exactly remember why I limited the fix to void type only...

The problem in your example also comes up in simpler dtypes, eg

    >>> dtype(('u4', 'u2,u2'))
    dtype([('f0', '<u2'), ('f1', '<u2')])
    >>> dtype(('u4', 'u2,u2')).type
    numpy.uint32

note how the first line does not show the 'u4' part.

As a sidenote, your dtype of `[("A", ">u4", [("comps", '2>u2')])]` surprised me a little. I would have expected that not to work because an extra `()` was needed, like `[("A", (">u4", [("comps", '2>u2')]))]`.
